### DISCUSSION TITLE
How does the updatedCallback(model, previousState) method work?

### DISCUSSION THREAD

**Discussion by TheMouster** [defx/synergy#10] [Q&A] [answered]
Ciao,

Me, again! I’ve made a currency formatter component that I update from a function call. However, while the attribute is updated the rendered value is not. Can someone (yeah, I know there’s only 2 of us here ;-)) show me an example of how the updatedCallback(model, previousState) method works? I’ve tried a couple of examples, but it’s not working for me. Any clues would be helpful.

Thanks

**Comment by TheMouster** [comment]
🆗,

After 2 hours of pushing pixels I have come up with the following, which works but only intermittently.

```
import { define } from 'https://unpkg.com/synergy@4.0.0';

class MoneyFormatter {
    formatted = "";
    value = 0.00;

    _formatMoney = (value) => parseFloat(value).toFixed(2).toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");

    updatedCallback(model, previousState) {
        this.formatted = this._formatMoney(model.money);
    }

    constructor(amount) {
        this.formatted = this._formatMoney(amount);
    }
}

const currencyValueFactory = (money) => {
    let dinero = new MoneyFormatter(money.money);

    return dinero;
};

define('currency-value', currencyValueFactory, "${{ dinero.formatted }}");
```

The element is rendered like so;

```
<currency-value money="0" id="curTotal"></currency-value>
```

… and has it’s money attribute updated by the results of a call to an API that returns the money value to many decimal places hence the need to format it.

The main question that comes out of my effort is why does the component not render at the time that the money attribute is changed? It is sometimes instant, which is great, but on other times it lags a couple of steps behind i.e. I’ve recomputed the value and it’s only just displaying the value as it was 3 changes ago. 

This lag makes me very sad.

**Reply by TheMouster** [reply]
BTW: I’ve had the lag thing happen previously in a text-area character counter I created, couldn’t figure it out then either.

**Comment by defx** [comment, marked as answer]
Hi @TheMouster , Synergy is designed to work with factory functions and plain objects instead of classes. This is how you can achieve the same thing...
```html
    <currency-value money="0" id="curTotal"></currency-value>
    <script type="module">
      import { define } from 'https://unpkg.com/synergy@5.0.0';

      const formatMoney = (value) =>
        parseFloat(value)
          .toFixed(2)
          .toString()
          .replace(/\B(?=(\d{3})+(?!\d))/g, ',');

      const currencyValueFactory = ({ money }) => {
        return {
          money,
          get formatted() {
            return formatMoney(this.money);
          },
        };
      };

      define('currency-value', currencyValueFactory, '${{ dinero.formatted }}');

      setTimeout(() => {
        document.getElementById('curTotal').setAttribute('money', 5.0);
      }, 500);
    </script>
```
A couple of things to note:
1) Synergy is now at v5!
2) Getters are perfect for deriving a new value from one or more other values on your model.

**Reply by defx** [reply]
`updatedCallback` is a lifecycle method that is invoked after each DOM update. I added this method for handling side effects such as data persistence, and it shouldn't be used to update the model itself. The `previousState` is provided so that you can make your side effect(s) conditional based on comparison of old and new values.

**Reply by TheMouster** [reply]
Cheers defx! Works a treat.
